Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php

```diff
@@ -40,16 +45,40 @@ public function publishEvents(ContentRepository $contentRepository, EventsToPubl
         $normalizedEvents = Events::fromArray(
             $eventsToPublish->events->map($this->eventNormalizer->normalize(...))
         );
-        $this->eventStore->commit(
+        $commitResult = $this->eventStore->commit(
             $eventsToPublish->streamName,
             $normalizedEvents,
             $eventsToPublish->expectedVersion
         );
+        $hooks = $this->hooksFactory->build($contentRepository);
+        $lock = $this->lockFactory->createLock($contentRepository->id->value . '_catchup');
+        $lock->acquire(true);
```

Comment:
detect if the lock was required from the same process (i.e. process is owner of the lock) and fail immediately if that's the case à la:
> You must not produce events (or send commands) in catch up hooks (except in the afterCatchUp implementation)

See line 80/81